Add advancing front surface reconstruction algorithm

Levi-Armstrong opened the pull request:
A implementation of a advancing front surface reconstruction algorithm based on the following papers.

**Advancing Front Papers**

1. High-Quality Extraction of Isosurfaces from Regular and Irregular Grids 
J. Schreiner, C. Scheidegger, C. Silva
IEEE Transactions on Visualization and Computer Graphics (Proceedings of IEEE Visualization); 2006

1. Direct (Re)Meshing for Efficient Surface Processing 
J. Schreiner, C. Scheidegger, S. Fleishman, C. Silva 
Computer Graphics Forum (Proceedings of Eurographics); 2006 

1. Triangulating Point-Set Surfaces With Bounded Error 
C. Scheidegger, S. Fleishman, C. Silva 
Proceedings of the third Eurographics/ACM Symposium on Geometry Processing; 2005 

**Testing**
`pcl_advancing_front_reconstruction bun0.pcd bun0.ply -radius 0.05 -rho 0.6`

**Debug interface**
In cmake gui enable BUILD_surface_afront_debug and rebuild. When running the command above a pcl viewer window will open and press numbers 0-9 to generated 10^x triangles.

**Maintainers**
This includes commits from the PR #1960. I planned on waiting for the PR to be merged before submitting this, but my budget goes away at the end of this month so I wanted to get this submitted before then.

taketwo commented:
Hi, thanks again for contributing. This PR is particularly hard to deal with: a complex algorithm implemented in some 3.5K lines of code. I'm no expert in meshing in general and this algorithm in particular, so I won't be able to assess this code in terms of how well it implements the papers. Further, even thoroughly checking the logic and exploring refactoring possibilities is well beyond what I can realistically do. Therefore, in my review I will concentrate on making sure that the overall code style is in the PCL spirit.

So far I have only scanned the code. Here are initial suggestions:
* Class name. I'd prefer to spell out the "advanced" completely.
* File names. Should match class name, but in `snake_case`, of course.
* License. It's alright to have your employer in the copyright, but please also add a line for Open Perception.
* Debugging visualization. I can imagine it was quite helpful during development and I acknowledge that it is incredibly instructive for understanding how the algorithm works. However, I think we should not include it in the class, even if carefully isolated with macro guards.

My next step will be reviewing the `advancing_front_utils.h` file. Some of the functions there are already present in PCL, some may be handy for other users and thus may be moved to the `common` module. So, ideally, I would like to see this file dissolving.

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+  {
+
+    /** \brief Align pn's normal with av so they point in the same direction */
+    bool
```
Do we need return type? I don't see you ever using it.

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+
+    /** \brief Align pn's normal with av so they point in the same direction */
+    bool
+    alignNormal (Eigen::Vector3f &pn, const Eigen::Vector3f &av)
```
Maybe `alignNormal`__s__?

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+    bool
+    alignNormal (Eigen::Vector3f &pn, const Eigen::Vector3f &av)
+    {
+      double dot = pn.dot (av);
```
No need to upcast to `double`. And in general, this function can be reduced to:

```cpp
if (pn.dot (av) < 0.0f)
  pn *= -1;
```

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+    bool
+    alignNormal (AfrontVertexPointType &pn, const AfrontVertexPointType &av)
+    {
+      Eigen::Vector3f normal = pn.getNormalVector3fMap ();
```
I think it should work if you pass `pn.getNormalVector3fMap ()` directly to the `alignNormal` function, the underlying data should be modified.

I'd move these two functions and the next one (`checkNormal`) to the `common` module. I don't really see any existing header file there that would naturally fit them, so propose to create a new one "normals.h".

While doing so, I'd make `alignNormals` templated on the point types, something like:

```cpp
template <typename PointNT1, typename PointNT2>
alignNormals (PointNT1& pn, const PointNT2& av)
```

to ensure reusability.

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+
+    /** \brief Check if two normals are within a tolerance */
+    bool
+    checkNormal (const Eigen::Vector3f n1, const Eigen::Vector3f n2, const double &tol)
```
Perhaps `checkNormalsEqual`?
`tol` -> `angular_threshold`
It also makes sense to mention in the docstring that the threshold is in radians.

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+        return false;
+
+      double denom = n1.norm () * n2.norm ();
+      double angle = acos (dot / denom);
```
Please use `std::acos`. Also, this function allows denormalized normals. What if one of the norms is zero?

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+
+      double denom = n1.norm () * n2.norm ();
+      double angle = acos (dot / denom);
+      if (angle > tol)
```
`return angle <= tol`;

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+
+    /** \brief Convert Eigen Vector3f to PCL PointXYZ */
+    pcl::PointXYZ
+    convertEigenToPCL (const Eigen::Vector3f &p)
```
Only used in debug branch code, so we can remove them as well.

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+    }
+
+    AfrontVertexPointType
+    convertPointNormalToAfrontPointType (const pcl::PointNormal &p)
```
Didn't find any usages of this one.

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+    }
+
+    AfrontGuidanceFieldPointType
+    convertAfrontPointTypeToAfrontGuidanceFieldPointType (const AfrontVertexPointType &p, const double rho)
```
Only used in one place, so may be moved directly there. Alternatively, make this a member function of the vertex point type.

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+      * \return The mid point of the half edge
+      */
+    Eigen::Vector3f
+    getMidPoint (const Eigen::Vector3f &p1, const Eigen::Vector3f &p2)
```
This is really trivial and is only used in two places, I'd simply expand this at the points of usage.

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+      * \return DistPoint2LineResults
+      */
+    DistPoint2LineResults
+    distPoint2Line (const Eigen::Vector3f &p1, const Eigen::Vector3f &p2, const Eigen::Vector3f &p)
```
There is a point to line distance function in `common` in "distances.h" can we reuse it?

Levi-Armstrong replied:
It looks like they are for checking against a line with infinite length, where I need to check against a line segment. Should I add these to this file? I would rename the methods to:

distPointToLineSeg
distLineSegToLineSeg

taketwo replied:
Yes, please. I'd spell "segment" completely, and also move "distance" part to the end to be similar with already existing function names, i.e. `pointToLineSegmentDistance` and `lineSegmentToLineSegmentDistance`. What do you think?

Levi-Armstrong replied:
Sounds good.

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+      * \return DistLine2LineResults
+      */
+    DistLine2LineResults
+    distLine2Line (const Eigen::Vector3f &p1, const Eigen::Vector3f &p2, const Eigen::Vector3f &p3, const Eigen::Vector3f &p4)
```
Can we integrate this into aforementioned header file?

taketwo commented on surface/include/pcl/surface/advancing_front_utils.h:
```diff
@@ -0,0 +1,310 @@
+      * \return IntersectionLine2PlaneResults
+      */
+    IntersectionLine2PlaneResults
+    intersectionLine2Plane (const Eigen::Vector3f &p1, const Eigen::Vector3f &p2, const Eigen::Vector3f &origin, const Eigen::Vector3f &u, const Eigen::Vector3f &v)
```
There is a function to compute line with plane intersection in `common` module "intersections.h" file. Can we reuse it?

Levi-Armstrong replied:
I only see lineWithLineIntersection and planeWithPlaneIntersection but no lineWithPlaneIntersection. Is it located somewhere else?

taketwo replied:
Sorry, my bad. Can we add your function there, then?

Levi-Armstrong replied:
OK good. I will move the function to the intersection.h file.

SergioRAgostinho commented:
#1960 was merged so let's rebase this one to remove the redundant commits.

moriarty commented:
Hey @taketwo, I just saw this mentioned in a talk at ROSCon, I agree with your first comment- this is hard to review.

I just thought I'd mention the ROSCon talk video should be available either now or soon.

taketwo commented:
Hi Alex, after having read your comment several times, I'm still not sure what was mentioned in the ROSCon talk and what relation it has with this PR. But I'm intrigued, please post a link when it's online ;)

moriarty commented:
Yes, I was in the talk when I commented above.
I'll post a link when it's online, it was available via live stream, but now we wait for Ubuntu to upload the videos and they expect it to take a month.

The talk gave a bit of a explaination, showed some results and then mentioned that it would be available soon as the PR was being reviewed. The results looked good and useful so I checked it out during the talk.

@taketwo, I saw your comment that it's hard to check the algorithm, and that you'd just review if the code was in the style of PCL, so I thought it would be helpful to see the results. Maybe @Levi-Armstrong can link to results somewhere.

Levi-Armstrong commented:
Sure, the results I presented are shown [here](http://www.pcl-users.org/Add-New-Advancing-Front-Mesher-td4044768.html). I am currently traveling back from ROSCon 2017 and on Monday I will begin addressing the comments above.

SergioRAgostinho commented:
FYI: I'm starting to look at this again.

SergioRAgostinho commented on surface/CMakeLists.txt:
```diff
@@ -151,14 +163,20 @@ if(build)
+    else(BUILD_surface_afront_debug)
+      target_link_libraries("${LIB_NAME}" pcl_common pcl_search pcl_kdtree pcl_octree ${VTK_LIBRARIES} ${ON_NURBS_LIBRARIES})
+    endif(BUILD_surface_afront_debug)
+
```
Do you see how things are done for QHull in line 180, that's how you should link with VTK. Move your condition below line 194.

SergioRAgostinho commented on surface/CMakeLists.txt:
```diff
@@ -1,6 +1,10 @@
+  set(SUBSYS_DEPS common search kdtree octree geometry visualization)
+else(BUILD_surface_afront_debug)
+  set(SUBSYS_DEPS common search kdtree octree geometry)
+endif(BUILD_surface_afront_debug)
```
This is the perfect file to define your `BUILD_surface_afront_debug` option. But this block needs to move below line 15, since if the surface module is not build, this option should not pop-up.

You're adding a dependency to `geometry` which in turn only depends of `common`. `geometry` has a number of meshing primitives so I would say that functionally this dependency makes sense. To be checked is exactly what you use from there.

Your conditional branch should not set SUBSYS_DEPS directly, it should just append the visualization dependency to it. I want the module mandatory dependencies to appear on top first, isolated from whatever options the user decides later down the road, in case the `surface`module is built

SergioRAgostinho commented on test/surface/CMakeLists.txt:
```diff
@@ -10,6 +10,10 @@ PCL_SUBSYS_DEPEND(build "${SUBSYS_NAME}" DEPS ${SUBSYS_DEPS} OPT_DEPS ${OPT_DEPS
   if (BUILD_io AND BUILD_features)
+    IF(BUILD_surface_afront_debug)
+        add_definitions(-DAFRONTDEBUG)
+    ENDIF(BUILD_surface_afront_debug)
```
Keeping in mind my comment of adding definitions, you should not need to add this here as well. It should come exported the moment you link with the `surface` target.

SergioRAgostinho commented on tools/CMakeLists.txt:
```diff
@@ -8,6 +8,9 @@ PCL_SUBSYS_OPTION (BUILD_tools "${SUBSYS_NAME}" "${SUBSYS_DESC}" ${DEFAULT} "${R
 if (BUILD_tools)
+  IF(BUILD_surface_afront_debug)
+      add_definitions(-DAFRONTDEBUG)
+  ENDIF(BUILD_surface_afront_debug)
```
Same comment as before

Levi-Armstrong commented on surface/include/pcl/surface/advancing_front.h:
```diff
@@ -0,0 +1,723 @@
+#ifndef PCL_SURFACE_ADVANCING_FRONT_H_
+#define PCL_SURFACE_ADVANCING_FRONT_H_
+
+#define PCL_NO_PRECOMPILE
```
I believe it was left over from when I was developing outside pcl. I was getting errors related to my custom point type when using them with other libraries like mls and kdtree, but when I added the define they all went away.

SergioRAgostinho replied:
Any reason for not registering your point type in `point_types.h[pp]` like all the others? Looking at the points included there I would say yours is a perfectly eligible candidate.

taketwo replied:
This is a support point type for this specific algorithm. What use will it have outside?
Added deficit: will increase build time and binary sizes due to additional template instantiations for all PCL algorithms.

Levi-Armstrong replied:
The plan was to only be used by this algorithm. I need to cache additional information for each point in the mesh to support the mesher, but there should not be any use for it outside of this.

SergioRAgostinho replied:
:+1: makes sense. Let's keep it then.

Levi-Armstrong commented:
FYI: I should be able to start addressing the open items this week.

SergioRAgostinho commented:
Thanks. Let's address the things by stages, starting with just CMake related issues at first. If possible leave each review stage on it's own commit to make it easier to traverse the history of changes.

taketwo commented:
Some of the CMake issues are related to debug visualization. As I wrote early on, I think we should just get rid of it. What's your stance on this?

SergioRAgostinho commented:
I was wondering if we could refactor that code and use it as a tool. I had a "similar" situation in the past, when I wrote a tracker class, and (of course) found out at some point that it was necessary to inject some more code to aid in understanding what was happening with the tracker. My solution for that was to derive the tracker debugger as a child class of the tracker. The debugger was working as the middle man, wrapping all calls to the parent while having access to the protected members. The debugger class would only reside in the tools module, alongside an application which invokes it. I was wondering if such strategy could be applied here as well.

taketwo commented:
This sounds like a reasonable thing to do, although will require extra effort from Levi.

Levi-Armstrong commented:
I will see what I can do. I believe this would be good to have since the paper mentions advanced techniques for reconstructing sharp edges where the tool would be useful for implementing this in the future.

SergioRAgostinho commented:
Thanks! Then I would suggest to pull the visualization code/tool for another PR, and for now strip this one of everything related to that. Does it sound reasonable? I'll try to go through another review stage in the upcoming days.

SergioRAgostinho commented on surface/include/pcl/surface/advancing_front_point_type.h:
```diff
@@ -0,0 +1,174 @@
+ *
+ * Point Cloud Library (PCL) - www.pointclouds.org
+ * Copyright (c) 2017-, Southwest Research Institute
+ *
```
I think Sergey already mentioned this somewhere else but you need to add Open Perception here as well.

SergioRAgostinho commented on surface/include/pcl/surface/advancing_front_point_type.h:
```diff
@@ -0,0 +1,174 @@
+#define PCL_SURFACE_ADVANCING_FRONT_POINT_TYPE_H_
+
+#include <pcl/point_types.h>
+#include <pcl/register_point_struct.h>
```
This one should not be required, since it's already included by `point_types.h`

SergioRAgostinho commented on surface/include/pcl/surface/advancing_front_point_type.h:
```diff
@@ -0,0 +1,174 @@
+
+#include <pcl/point_types.h>
+#include <pcl/register_point_struct.h>
+#include <pcl/impl/instantiate.hpp>
```
Why is this required here? There's nothing on this file which makes instantiations.

SergioRAgostinho commented on surface/include/pcl/surface/advancing_front_point_type.h:
```diff
@@ -0,0 +1,174 @@
+
+namespace pcl
+{
+  namespace afront
```
I don't think this extra namespace is required. You're already naming everything `Afront...`.

SergioRAgostinho commented on surface/include/pcl/surface/advancing_front_point_type.h:
```diff
@@ -0,0 +1,174 @@
+        curvature = p.curvature;
+        max_step = p.max_step;
+        max_step_search_radius = p.max_step_search_radius;
+      }
```
Why are you required to define this? This looks like the default copy constructor and the compiler should generate one for you automatically.

Levi-Armstrong replied:
I just noticed that most of the existing if not all did this so I assumed that there was a reason behind it. Let me know if you would like it removed?

SergioRAgostinho replied:
Yes please. :+1: There's a couple of  things that happen if you define the copy constructor, because intuitively you're telling the compiler "Wait. My copy process is more complex than usual" and the compiler won't generate a couple of useful methods for you in that case. See https://stackoverflow.com/questions/3734247/what-are-all-the-member-functions-created-by-compiler-for-a-class-does-that-hap#3734268 for more info.

SergioRAgostinho commented on surface/include/pcl/surface/advancing_front_point_type.h:
```diff
@@ -0,0 +1,174 @@
+      friend std::ostream &
+      operator<< (std::ostream &os, const AfrontVertexPointType &p)
+      {
+        os << p.x << "\t" << p.y << "\t" << p.z;
```
I think you should actually print all your point types internal members.

SergioRAgostinho commented on surface/include/pcl/surface/advancing_front_point_type.h:
```diff
@@ -0,0 +1,174 @@
+        data_n[3] = 0.0f;
+        curvature = p.curvature;
+        ideal_edge_length = p.ideal_edge_length;
+      }
```
Same comment as before.

SergioRAgostinho commented on surface/include/pcl/surface/advancing_front_point_type.h:
```diff
@@ -0,0 +1,174 @@
+      friend std::ostream &
+      operator<< (std::ostream &os, const AfrontGuidanceFieldPointType &p)
+      {
+        os << p.x << "\t" << p.y << "\t" << p.z;
```
Same comment as before.

SergioRAgostinho commented on surface/include/pcl/surface/advancing_front_point_type.h:
```diff
@@ -0,0 +1,174 @@
+  } // namespace afront
+} // namespace pcl
+POINT_CLOUD_REGISTER_POINT_STRUCT (pcl::afront::AfrontVertexPointType,
+                                   (float, x, x) (float, y, y) (float, z, z) (float, normal_x, normal_x) (float, normal_y, normal_y) (float, normal_z, normal_z) (float, curvature, curvature) (float, max_step, max_step) (float, max_step_search_radius, max_step_search_radius))
```
Can you span this over multiple lines to increase readability? Each field on each line.

SergioRAgostinho commented on surface/include/pcl/surface/advancing_front_point_type.h:
```diff
@@ -0,0 +1,174 @@
+                                   (float, x, x) (float, y, y) (float, z, z) (float, normal_x, normal_x) (float, normal_y, normal_y) (float, normal_z, normal_z) (float, curvature, curvature) (float, max_step, max_step) (float, max_step_search_radius, max_step_search_radius))
+
+POINT_CLOUD_REGISTER_POINT_STRUCT (pcl::afront::AfrontGuidanceFieldPointType,
+                                   (float, x, x) (float, y, y) (float, z, z) (float, normal_x, normal_x) (float, normal_y, normal_y) (float, normal_z, normal_z) (float, curvature, curvature) (float, ideal_edge_length, ideal_edge_length))
```
Same comment as above.

SergioRAgostinho commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+    double d2121 = (p2 - p1).dot ((p2 - p1));
+    double d4321 = (p4 - p3).dot ((p2 - p1));
+    double d1343 = (p1 - p3).dot ((p4 - p3));
+    double d4343 = (p4 - p3).dot ((p4 - p3));
```
All these need to be switched to float precision and set to const.

SergioRAgostinho commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+    double d1343 = (p1 - p3).dot ((p4 - p3));
+    double d4343 = (p4 - p3).dot ((p4 - p3));
+
+    double denom = d2121 * d4343 - d4321 * d4321;
```
This is the determinant of the linear system of equations you have to solve and according to my equations this should be
```cpp
const float det = - d2121 * d4343 + d4321 * d4321;
```

SergioRAgostinho commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+    double d4343 = (p4 - p3).dot ((p4 - p3));
+
+    double denom = d2121 * d4343 - d4321 * d4321;
+    if (denom < 1.0e-8) // The lines are parallel
```
You need to check the absolute value in case the determinant is negative

SergioRAgostinho commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+      result.parallel = true;
+
+      dist = pointToLineSegmentDistance (p1, p2, p3);
+      result.mu[0] = dist.mu;
```
You only need to compute the point to line segment once because they're parallel. In this situation, you should set both `mu`s and the corresponding interception points to NaN

SergioRAgostinho commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+    else
+    {
+      result.mu[0] = (d1343 * d4321 - d1321 * d4343) / denom;
+      result.mu[1] = (d1343 + result.mu[0] * d4321) / d4343;
```
According to my calculations this should be 
```cpp
mu0 =  d4343 * d1321 - d4321* d1343;
mu1 =  d4321 * d1321 - d2121* d1343;
```

SergioRAgostinho commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+      if (result.mu[1] > 1.0)
+        result.mu[1] = 1.0;
+      else if (result.mu[1] < 0.0)
+        result.mu[1] = 0.0;
```
Just like before, these clampings make no sense.

SergioRAgostinho replied:
Use the combination of `std::min` and `std::max`

SergioRAgostinho commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+        result.mu[1] = 0.0;
+
+      result.p[0] = p1 + result.mu[0] * (p2 - p1);
+      result.p[1] = p3 + result.mu[1] * (p4 - p3);
```
This makes no sense if the `mu`s are clamped.

SergioRAgostinho commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+
+    result.d = (result.p[1] - result.p[0]).norm ();
+    return result;
+  }
```
Similar comment here as before in `pointToLineSegmentDistance`. Move the function to a cpp file, change the signature in the same way and make the corresponding overload.

Levi-Armstrong commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+    else if (c2 <= c1)
+      results.mu = 1.0;
+    else
+      results.mu = c1 / c2;
```
It is measuring the distance between a point and a Line segment. Example for the case of triangle edge, I need to know the distance from a point to a line segment so mu must be clamped between 0 and 1 to get the correct distance.

SergioRAgostinho replied:
Yes, but this function will be used for more than your triangular case, so that check needs to be done a posteriori.

Levi-Armstrong replied:
That's fine as long as the function return mu I can check it in my code.

SergioRAgostinho replied:
Then again, the name of the function is distance to line segment and not really distance to line. 

@taketwo opinion: Should we allow the option of clamping as a parameter?

Levi-Armstrong replied:
Another option could be to modify the existing sqrPointToLineDistance to calculate mu. Then I can get the information I need if that simplifies things.

SergioRAgostinho replied:
Oh true, that methods exists! Then my comment makes no sense. I'll delete/strike through everything which is not longer valid.

SergioRAgostinho replied:
For clamping please use a combination of `std::max` and `std::min`

Levi-Armstrong commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+      results.mu = c1 / c2;
+
+    results.p = results.points[0] + results.mu * v;
+    results.d = (p - results.p).norm ();
```
Explanation above.

Levi-Armstrong commented on common/include/pcl/common/distances.h:
```diff
@@ -197,6 +197,125 @@ namespace pcl
+    float mu[2];               /**< \brief The corresponding line paramentric value resulting in the minimum distance. */
+    double d;                  /**< \brief The distance between both line segments */
+    bool parallel;             /**< \brief Indicate that both lines are parallel. */
+  };
```
Also I could switch to using this function lineToLineSegment possibly but may need to make a change.

SergioRAgostinho replied:
Let's keep your functions specialized for line segments which are finite. Let's make that clear in the function description, so that (contrary to me) people understand that this could mean that this distance is not necessarily measure orthogonal to the line segments.

SergioRAgostinho commented:
Update the comments. After the discussion, the target now is only changes in the functions signatures and the moving to a normal .cpp file.

Levi-Armstrong commented:
Sounds good! I will make the changes for the .cpp and then wait for your response on how to update the function signature.

SergioRAgostinho commented:
After you called my attention to the fact that `lineToLineSegment` and `sqrPointToLineDistance` exist, it stopped making sense to me let the user decide if he wants to clamp or not in the function signature of `pointToLineSegmentDistance` and `lineSegmentToLineSegmentDistance`, because these other methods do new methods provide that functionality.

I'm expecting signatures like this

```cpp
float
pcl::pointToLineSegmentDistance (const Eigen::Vector3f &p1,
                                 const Eigen::Vector3f &p2,
                                 const Eigen::Vector3f &p,
                                 Eigen::Vector3f &pi
                                 float &mu);


float
pcl::pointToLineSegmentDistance (const Eigen::Vector3f &p1,
                                 const Eigen::Vector3f &p2,
                                 const Eigen::Vector3f &p);

float
pcl::lineSegmentToLineSegmentDistance (const Eigen::Vector3f &p1,
                                       const Eigen::Vector3f &p2,
                                       const Eigen::Vector3f &p3,
                                       const Eigen::Vector3f &p4,
                                       Eigen::Vector3f &pi1,
                                       float &mu1,
                                       Eigen::Vector3f &pi2,
                                       float &mu2);


float
pcl::lineSegmentToLineSegmentDistance (const Eigen::Vector3f &p1,
                                       const Eigen::Vector3f &p2,
                                       const Eigen::Vector3f &p3,
                                       const Eigen::Vector3f &p4);
```

Levi-Armstrong commented:
I should have time over the next few weeks to address the remaining changes.

Micalson commented:
@Levi-Armstrong Hi, I find the pull request about "A implementation of a advancing front surface reconstruction algorithm based on the following papers." But I cannot find the function in pcl repo.
I want to use the advancing front surface reconstruction algorithm, where could I find it? Thank you for your reply.

Regards,

Levi-Armstrong commented:
@Micalson The only way to use it currently, is to build my branch in this PR to be able to leverage this meshing algorithm at them moment. I need to wrap up a few things before this can be merged, but I wanted to fix one know issue at the moment. It currently sometimes will wrap on its self. I believe I have a solution but I have not had time to implement and test. If you run into any issues let me know.

Micalson commented:
@Levi-Armstrong I get it. Thanks for your reply.
